### PR TITLE
Fix bug of first time binding

### PR DESCRIPTION
**Pull request by rita-g-sonarsource**
I found a nasty bug during QA: I used the existing binding information in order to fetch the configurations, which caused a bug when there was no existing binding information (first time bind).
Hence I had to move the code that calculates the new binding configuration, and use that new configuration.

### CODE REVIEW

File: src/Integration/Binding/BindingController.cs [hunk 1]
```diff
@@ -158,24 +158,20 @@ void IBindingWorkflowExecutor.BindProject(BindCommandArgs bindingArgs)
...
                 host,
-                bindingArgs.Connection,
-                bindingArgs.ProjectKey,
-                bindingArgs.ProjectName,
                 modeToBind,
                 host.Logger);
 
             var unboundProjectFinder = new UnboundProjectFinder(host, host.Logger);
 
-            var cSharpVBBindingConfigProvider = new CSharpVBBindingConfigProvider(host.SonarQubeService, nugetBindingOp, currentConfiguration, host.Logger);
```
Comment (rita-g-sonarsource) [inline src/Integration/Binding/BindingController.cs:169]: This was an oversight on my part -- all the calculation should be done based on _new configuration_, and not on the current one.

File: src/Integration/Binding/BindingProcessImpl.cs [hunk 2]
```diff
@@ -122,8 +127,10 @@ public async Task<bool> DownloadQualityProfileAsync(IProgress<FixedStepsProgress
                 }
                 this.InternalState.QualityProfiles[language] = qualityProfileInfo;
 
+                var bindingConfiguration = WriteBindingInformation();
```
Comment (rita-g-sonarsource) [inline src/Integration/Binding/BindingProcessImpl.cs:130]: The location of this calculation sucks. Logically it needs to be done after fetching the quality profile, and before calling GetConfigurationAsync.

File: src/Integration/Binding/BindingProcessImpl.cs [hunk 3]
```diff
@@ -143,6 +150,40 @@ public async Task<bool> DownloadQualityProfileAsync(IProgress<FixedStepsProgress
             return true;
         }
 
+        /// <summary>
+        /// Will add/edit the binding information for next time usage
+        /// </summary>
+        private BindingConfiguration WriteBindingInformation()
```
Comment (rita-g-sonarsource) [inline src/Integration/Binding/BindingProcessImpl.cs:156]: Cut-paste of the logic that was inside SolutionBindingOperation.

File: src/Integration/Binding/ISolutionBindingOperation.cs [hunk 4]
```diff
@@ -36,7 +34,7 @@ namespace SonarLint.VisualStudio.Integration.Binding
     /// </remarks>
     public interface ISolutionBindingOperation : ISolutionBindingConfigFileStore
     {
-        void Initialize(IEnumerable<Project> projects, IDictionary<Language, SonarQubeQualityProfile> profilesMap);
```
Comment (rita-g-sonarsource) [inline src/Integration/Binding/ISolutionBindingOperation.cs:39]: No longer needed, as it was only for the calculation of the BindingConfiguration.

File: src/Integration/NewConnectedMode/ConfigurationProvider.cs [hunk 5]
```diff
@@ -70,8 +70,10 @@ public bool Persist(BoundSonarQubeProject project, SonarLintMode bindingMode)
 
             var writeSettings = GetWriteSettings(bindingMode);
 
-            return writeSettings.HasValue && 
+            var success = writeSettings.HasValue && 
                    solutionBindingSerializer.Write(writeSettings?.ConfigPath, project, writeSettings?.OnSuccessfulFileWrite);
+
+            return success ? TryGetBindingConfiguration(writeSettings?.ConfigPath, project, bindingMode) : null;
```
Comment (rita-g-sonarsource) [inline src/Integration/NewConnectedMode/ConfigurationProvider.cs:76]: solutionBindingSerializer.Write only queues the writing operation, so I couldn't call GetConfiguration. I had to instead use the created `project` and initialize a new BindingConfiguration object.

File: src/Integration/Binding/BindingProcessImpl.cs [hunk 6]
```diff
@@ -143,6 +150,40 @@ public async Task<bool> DownloadQualityProfileAsync(IProgress<FixedStepsProgress
             return true;
         }
 
+        /// <summary>
```
Comment (duncanp-sonar) [inline src/Integration/Binding/BindingProcessImpl.cs:153]: Please call out in a review comment when you've moved a chunk of code from one class to another and whether you've made any changes to it - it's hard to tell in the GitHub web UI.

File: src/Integration/NewConnectedMode/ConfigurationProvider.cs [hunk 7]
```diff
@@ -82,7 +84,12 @@ private BindingConfiguration TryGetBindingConfiguration(string bindingPath, Sona
             }
 
             var boundProject = solutionBindingSerializer.Read(bindingPath);
+            
+            return TryGetBindingConfiguration(bindingPath, boundProject, sonarLintMode);
+        }
 
+        private BindingConfiguration TryGetBindingConfiguration(string bindingPath, BoundSonarQubeProject boundProject, SonarLintMode sonarLintMode)
```
Comment (duncanp-sonar) [inline src/Integration/NewConnectedMode/ConfigurationProvider.cs:91]: Rename to `Create`

File: src/Integration/NewConnectedMode/ConfigurationProvider.cs (hunk 5 above)
Comment (duncanp-sonar) [reply src/Integration/NewConnectedMode/ConfigurationProvider.cs:76]: As discussed offline, I think that's ok. We need to create a new instance somewhere.

**Pull request merged**